### ISSUE TITLE
A few questions related to training

### ISSUE THREAD

**Issue by dongzhang04** [vitusbenson/greenearthnet#3]
Hi Vitus.

Firstly, just a couple questions about training Contextformer. We're wondering how long the training took you and how many GPUs you used. We're also curious about what you think the minimum number of training sites and timesteps is and whether we could use irregular time intervals rather than specifically 5 day intervals between timesteps. 

Secondly, I've been retraining Contextformer on 10 of the training sites to confirm that we are able to retrain the model, however I've been encountering this warning message. 

<img width="1453" height="118" alt="Image" src="https://github.com/user-attachments/assets/a3be18d3-0a64-4db9-85e8-b99588fb9ddb" />

I also wanted to confirm whether the metric is correct. At first it was looking for a 'veg_score' metric but I had modified earthnet-models-pytorch/data/init.py file to use to the 'RMSE_Veg' metric that was specified in this initial config output (which would print when using 'veg_score' as well). 

<img width="621" height="239" alt="Image" src="https://github.com/user-attachments/assets/9cd5824f-9713-45b1-853c-547634d3ea03" />

Thanks!

**Comment by vitusbenson** [comment]
Hi Dong,

apologies for the delayed response.

> We're wondering how long the training took you and how many GPUs you used. 

For the final Contextformer, I think I trained with 2 A100 GPUs for ~2 days (but not sure tbh).

One more comment: I see your training seems pretty slow. This can be quite tricky to get right, but some tips:
- try to store all data on the local NVMe SSD of the GPU machine for fast file access
- Change the `num_workers` argument and check which one gives fastest access
- If you are bold you could also try to store the dataset in the `fast_access` format, the dataloading code for this is still in the dataloader - but how to rewrite the dataset in that format you'd have to infer yourself

> We're also curious about what you think the minimum number of training sites and timesteps is

The more the merier, I think GreenEarthNet uses around 20k training "sites", as in minicubes. You can definitely adjust the timesteps.. the design choice of 10 context & 20 prediction could probably be optimized.

> use irregular time intervals rather than specifically 5 day intervals between timesteps

In theory for sure, but you'd have to adjust quite some code. The whole `forward` of Contextformer (https://github.com/earthnet2021/earthnet-models-pytorch/blob/f4a12cc44e9882c6621bc0e86e4639c6ee9c84de/earthnet_models_pytorch/model/contextformer.py#L326) is currently written such that it only accepts 5-daily S-2 images. You'd have to generalize that and also probably also adjust the temporal positional encoding (https://github.com/earthnet2021/earthnet-models-pytorch/blob/f4a12cc44e9882c6621bc0e86e4639c6ee9c84de/earthnet_models_pytorch/model/contextformer.py#L499-L504) accordingly.

> Secondly, I've been retraining Contextformer on 10 of the training sites to confirm that we are able to retrain the model, however I've been encountering this warning message.

Mmh this is hard to debug for me. Can you share the config file that you use and also put the `earthnet-models-pytorch` changes somewhere?
Otherwise you could also try to debug yourself, i'd probably check if:
1) https://github.com/earthnet2021/earthnet-models-pytorch/blob/f4a12cc44e9882c6621bc0e86e4639c6ee9c84de/earthnet_models_pytorch/metric/rmse.py#L112 is ever called
2) This should in principle be here: https://github.com/earthnet2021/earthnet-models-pytorch/blob/f4a12cc44e9882c6621bc0e86e4639c6ee9c84de/earthnet_models_pytorch/task/workflow.py#L245 - but maybe something broke with upgrading to newer pytorch lightning.

**Comment by dongzhang04** [comment]
Hi Vitus,

sorry for my late response as well, I've been sick recently and overlooked the reply.

Once again, thanks for the detailed responses to all our questions. 

With regards to the warning, the changes to earthnet-models-pytorch can be found at [https://github.com/dongzhang04/greenearthnet/blob/9e8467d2404fd528d16505e690b342e6e4aca2b4/development/greenearthnet/training/__init__.py#L50-L59](https://github.com/dongzhang04/greenearthnet/blob/9e8467d2404fd528d16505e690b342e6e4aca2b4/development/greenearthnet/training/__init__.py#L50-L59)

and the config file I used was:

```yaml
Architecture: "contextformer"

Seed: 42

Setting: greenearthnet

Logger:
    save_dir: "experiments/"

Checkpointer:
    save_top_k: 1
    save_last: True
    every_n_epochs: 1

Trainer:
    # gpus: 2
    # strategy: 'ddp'
    accelerator: 'cpu'
    strategy: 'single_device'
    log_every_n_steps: 32
    max_epochs: 100
    gradient_clip_val: 1

Data:
    test_track: "ood-t_chopped"
    train_batch_size: 16
    val_batch_size: 8
    test_batch_size: 8
    num_workers: 32
    dl_cloudmask: True
    allow_fastaccess: False
    new_valset: True

Task:
    loss:
        name: "MaskedL2NDVILoss"
        lc_min: 10
        lc_max: 40
        ndvi_pred_idx: 0
        ndvi_targ_idx: 0
        pred_mask_value: -1
        scale_by_std: False
    context_length: 10
    target_length: 20
    n_stochastic_preds: 1
    optimization:
        optimizer:
            - 
                name: AdamW
                args: 
                    betas: [0.9, 0.999]
                lr_per_sample: 0.00000125
        lr_shedule:
            -
                name: MultiStepLR
                args:
                    milestones: [190] #[2, 20, 50, 90]
                    gamma: 0.1
    n_log_batches: 2
    compute_metric_on_test: True

Model:
    patch_size: 4
    n_image: 8
    n_weather: 24
    n_hidden: 256
    n_out: 1
    n_heads: 8
    depth: 3
    mlp_ratio: 4.0
    mtm: True
    leave_n_first: 3
    p_mtm: 0.7
    p_use_mtm: 0.5
    mask_clouds: True
    predict_delta: False
    predict_delta0: False
    predict_delta_avg: False
    predict_delta_max: False
    add_last_ndvi: True
    add_mean_ndvi: False
    pvt: True
    pvt_frozen: False
```

**Comment by dongzhang04** [comment]
Hi Vitus. 

Turns out the issue was that I'd missed the val_chopped dataset. I had been testing the training script with a subset of the training dataset and hadn't known to copy the validation data along with the training data. 

Apologies for the confusion.
